test/e2e/e2e_test.go

```diff
@@ -29,6 +30,136 @@ const (
 	exGwAnnotation       = "k8s.ovn.org/hybrid-overlay-external-gw"
 )
 
+// PodAnnotation describes the assigned network details for a single pod network. (The
+// actual annotation may include the equivalent of multiple PodAnnotations.)
+type PodAnnotation struct {
+	// IPs are the pod's assigned IP addresses/prefixes
+	IPs []*net.IPNet
+	// MAC is the pod's assigned MAC address
+	MAC net.HardwareAddr
+	// Gateways are the pod's gateway IP addresses; note that there may be
+	// fewer Gateways than IPs.
+	Gateways []net.IP
+	// Routes are additional routes to add to the pod's network namespace
+	Routes []PodRoute
+}
+
+// PodRoute describes any routes to be added to the pod's network namespace
+type PodRoute struct {
+	// Dest is the route destination
+	Dest *net.IPNet
+	// NextHop is the IP address of the next hop for traffic destined for Dest
+	NextHop net.IP
+}
+
+// Internal struct used to marshal PodAnnotation to the pod annotation
+type podAnnotation struct {
+	IPs      []string   `json:"ip_addresses"`
+	MAC      string     `json:"mac_address"`
+	Gateways []string   `json:"gateway_ips,omitempty"`
+	Routes   []podRoute `json:"routes,omitempty"`
+
+	IP      string `json:"ip_address,omitempty"`
+	Gateway string `json:"gateway_ip,omitempty"`
+}
+
+// Internal struct used to marshal PodRoute to the pod annotation
+type podRoute struct {
+	Dest    string `json:"dest"`
+	NextHop string `json:"nextHop"`
+}
+
+type annotationNotSetError struct {
+	msg string
+}
+
+func (anse annotationNotSetError) Error() string {
+	return anse.msg
+}
+
+// newAnnotationNotSetError returns an error for an annotation that is not set
+func newAnnotationNotSetError(format string, args ...interface{}) error {
+	return annotationNotSetError{msg: fmt.Sprintf(format, args...)}
+}
+
+// unmarshalPodAnnotation returns the default network info from pod.Annotations
+func unmarshalPodAnnotation(annotations map[string]string) (*PodAnnotation, error) {
```

Comment:
are  all these functions and structs taken from ovn-k8s package? is it possible to import them here or no?

Reply:
i tried, but ran into all sorts of weird import issues. Showed to @nerdalert and @aojea and we couldn't figure out, so we agreed to just copy it over to the tests.
Maybe land as is and we look at refactoring later on?

Reply:
this folder has its own go modules, it's independent from go-controller.
I don't know if you can play with the redirects in go.mod , but due to the amount of code I think that is easier to copy it directly, but I suggest to create a utils.go file in the same folder with the code, so we don't keep growing this files.

Reply:
aye, i can decouple on its own file

Reply:
sounds good

Reply:
done